mllib/src/main/scala/org/apache/spark/ml/feature/VectorSlicer.scala

```diff
@@ -110,22 +110,21 @@ final class VectorSlicer @Since("1.5.0") (@Since("1.5.0") override val uid: Stri
     }
 
     // Prepare output attributes
-    val inds = getSelectedFeatureIndices(dataset.schema)
-    val selectedAttrs = inputAttr.attributes.map { attrs =>
-      inds.map(index => attrs(index))
-    }
+    val selectedIndices = getSelectedFeatureIndices(dataset.schema)
+    val selectedAttrs = inputAttr.attributes.map { attrs => selectedIndices.map(attrs.apply) }
     val outputAttr = selectedAttrs match {
       case Some(attrs) => new AttributeGroup($(outputCol), attrs)
-      case None => new AttributeGroup($(outputCol), inds.length)
+      case None => new AttributeGroup($(outputCol), selectedIndices.length)
     }
 
-    // Select features
+    val sorted = selectedIndices.length > 1 && selectedIndices.sliding(2).forall(t => t(1) > t(0))
```

Comment:
Hm, I'm still not sure about this. Why make the caller do it and potentially get it wrong? just 'remember' internally whether it is sorted. That would optimize all calls to slice.

Reply:
> just 'remember' internally whether it is sorted.

I am not sure, but if we want to make it 'remember' internally whether it is sorted, I guess we need to add a new class `VectorSlicer(val indices: Array[Int]) { def slice(vec: Vector): Vector }`

Currently, `slice` is just a method of `SparseVector`, we may not make each `SparseVector` remember this sortedness?

Reply:
That would be the drawback, yeah, another boolean and a check every time one is created. The upside is all calls to slice are potentially optimized, and maybe, that sortedness can be used elsewhere?

Reply:
If we create an additional class (VectorSlicer) for this, I guess we can not use it elsewhere. Since this slicing indices can not be a member of vector.

Reply:
Not quite sure what you're proposing; I just have a moderately strong preference for not putting this logic in N callers if possible. OK, factoring it out to a utility class is a little better. Is the runtime check for sortedness just too expensive and/or not reusable in other functions?

Reply:
> I just have a moderately strong preference for not putting this logic in N callers if possible.

I am not sure how to impl it.

> Is the runtime check for sortedness just too expensive and/or not reusable in other functions?

Checking for sortedness maybe O(n), mayebe too expensive. The sortedness is only used in slicing for now.

Reply:
OK I trust your judgment. Let's refactor it if possible to make it simple for a caller and keep it private to Spark if possible.

Reply:
Yes. I think current impl maybe simple enough, and it is already private to spark now.

Reply:
Go ahead and merge as you see fit.